sycl/doc/extensions/DiscardQueueEvents/SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS.asciidoc

```diff
@@ -0,0 +1,223 @@
+_invalid_ event API behavior:
+[source,c++]
+----
+// the behavior is the same as for the default event, i.e. must throw an exception with the errc::invalid_object error code
+cl_event get() const
+
+// the behavior is the same as for the default event, i.e. must return true
+bool is_host() const;
+
+// the behavior is the same as for the default event, i.e. must return backend::host
+backend get_backend() const noexcept;
+
+// the behavior is the same as for the default event, i.e. must return a zero-sized vector
+std::vector<event> get_wait_list();
+
+// must throw an exception with the errc::invalid_object error code.
+void wait();
+
+// passing invalid event in the function must throw an exception with the errc::invalid_object error code.
+static void wait(const std::vector<event> &eventList);
+
+// must throw an exception with the errc::invalid_object error code.
+void wait_and_throw();
+
+// passing invalid event in the function must throw an exception with the errc::invalid_object error code.
+static void wait_and_throw(const std::vector<event> &eventList);
+
+// returns info::event_command_status::ext_oneapi_unknown status
+get_info<info::event::command_execution_status>() const;
+
+// the behavior is the same as for the default event, i.e. returns 0
+get_info<info::event::reference_count>() const;
+
+// must throw an exception with the errc::invalid_object error code.
+template <info::event_profiling param>
+typename param::return_type get_profiling_info() const;
+----
+
+The behavior when passing _invalid_ event to handler API:
+[source,c++]
+----
+// must throw an exception with the errc::invalid_object error code.
+handler::depends_on(event Event)
+
+// must throw an exception with the errc::invalid_object error code.
+handler::depends_on(const std::vector<event> &Events)
+----
+
+== Limitations
+
+The `discard_queue_events` property should be taken as a hint for SYCL implementation.
+In all kernel cases, SYCL implementation returns _invalid_ event, regardless of whether the optimization was done.
+Optimization occurs if the queue has `discard_queue_events` property and all features presented below are not met:
+
+ - Using out-of-order queue or using together with `enable_profiling` property
+ - Using https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/Assert/SYCL_ONEAPI_ASSERT.asciidoc[fallback assert feature]
+ - Using streams, buffer/image accessors (excluding local accessors)
+
+
+See details for each below:
+
+=== Using out-of-order queue and/or using together with enable_profiling property
+
+No optimization if a queue is created with the `discard_queue_events` property and
+the property list includes the `enable_profiling` property or does not include `in_order` property.
+
+=== Using fallback assert feature
+
+If a kernel that uses the fallback assert feature is submitted to a queue created with
+the `discard_queue_events` property, the implementation must throw an exception with the errc::invalid_object error code.

```

Comment:
Could you please tell why this is a "hard" limitation(throw an exception) rather than a "soft"(optimization doesn't work in this case) one?

Reply:
This is due to internal SYCL implementation. Since in that case, we need to somehow pass the information to the kernel launch function that the kernel uses the `assert` macro or not. I checked the code and found that if that limitation will be"soft" then we will have additional overhead(twice capture/release mutex `GlobalHandler::instance().getHandlerExtendedMembersMutex()`). --- this is due to a violation of our ABI policy as I cannot add a new field for CG and handler class.

But if we throw an exception as soon as we know that the `fallback assert feature` is enabled then we will not do this extra code for the case when `fallback assert feature` is disabled.

Moreover, fallback assert is not about performance, it has overhead even if the `assert` statement is not written in the kernel. And since the user who will use the new `SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS` extension clearly cares about performance, we can do it for fewer cases, but do it better/faster.

Reply:
Well, I looked at it again and found that we can make it "soft" and it will be inexpensive. I changed the doc.